Food Rework Update

whatston3 opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->

Update of #1159.  Cleanup done, use of flavorol, species-specific digstion in-progress.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->

Food rework's a long time coming.

## How to test
<!-- Describe the way it can be tested -->

TBD.  Eat stuff.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
TBD

dvir001 commented:
@whatston3 Is this ready for another round of final testing?
I honestly think it's all working fine for the first V.1 of the system.

After that we can add a lot more on top.

dvir001 reviewed (approved):
Honestly this is fine for 0.1v

GreaseMonk reviewed (approved):
i ate some food, i vomited some stuff, drank the vomit, tasted like nutrition
its good to go , just gotta gather some feedback after

ErhardSteinhauer commented on Resources/Prototypes/Body/Organs/arachnid.yml:
```diff
@@ -35,6 +35,13 @@
     sprite: Mobs/Species/Arachnid/organs.rsi
     state: stomach
   - type: Stomach
+    # specialDigestible:
+    #   tags:
+    #   - Fruit
+    #   - Pill
+    #   - BloodFood
+    #   - Crayon
+    #   - Soup
```
It's an upstream thing irc which is commented out because we don't use it

GreaseMonk replied:
I just see nothing on left hand side and it just looks like its added instead of commented

arimah replied:
Arachnids do not have a special diet on upstream, so these lines are in fact unique to us and should not be present: https://github.com/space-wizards/space-station-14/blob/master/Resources/Prototypes/Body/Organs/arachnid.yml

arimah commented on Resources/Prototypes/Body/Organs/moth.yml:
```diff
@@ -4,12 +4,16 @@
-    #   - Fruit # Frontier
-    #   - Pill # Frontier
-    #   - Crayon # Frontier
+    specialDigestible:
```
By allowing moths to eat only MothFood, this _severely_ restricts moth players. With no ability to eat fruit or pills, that leaves moths with like 4 foods. If there's no food ship around (think lowpop), this is considerably less than ideal. On upstream, moths can at least eat easily-obtained jumpsuits (which  I'm not advocating for here). Given the supremely limited selection of moth-friendly nommables, I think we should reconsider this. I don't think this will lead to better or more interesting roleplay; I think it'll lead to a bunch of starving moths.

arimah replied:
Based on a misunderstanding of how the new system works. Moths can eat regular foods (thanks to `components: - Food` below) _and_ MothFood. Disregard the above.

arimah commented on Resources/Prototypes/_NF/Entities/Objects/Consumable/Food/burger.yml:
```diff
@@ -44,3 +46,5 @@
     sprite: _NF/Objects/Consumable/Food/burger.rsi
     state: rosy-mothroach
+  - type: Food
+    quality: Nasty
```
Why is the regular mothroach burger high-quality while this is nasty? Shouldn't the rare and exotic version be even better?!

arimah reviewed (changes requested):
I like data. I collected data on every entity in game and put data in spreadsheet: https://docs.google.com/spreadsheets/d/1_VyrdJpWqONr_0vvlRyFE2a6PTKRV46HEtxyfTyQwo0/edit?usp=sharing
This data is based on every entity with a `FoodComponent` where `RequiresSpecialDigestion` is false *or* the entity contains a `MothFood` tag[^1]. The "Moth food?" column is true if the entity has the `MothFood` tag. The data is slightly poisoned by the sheer number of edible paper things, which I couldn't be bothered to filter out, and also includes produce that is only edible by cows (e.g. WheatBushel).

Some observations:
* These are perfectly safe to eat, despite being classified as toxins: FoodPieXeno, FoodPieXenoSlice, FoodBreadMeatSpider, FoodBreadMeatSpiderSlice. For comparison, FoodBurgerXeno is normal.
* The following dungeon foods should be classified as toxins: FoodNoodlesMeatballMoldy, FoodNoodlesMoldy, FoodRiceEggMoldy, FoodSaladCaesarMoldy, FoodSoupMeatballMoldy
* FoodBungoPit should probably be reclassified as a toxin, since it contains bungotoxin that will kill you in a hurry.
* ~~Moths are _really_ screwed over. 18 edible things, of which 10 are junk (9 of the junk things are PSBs). They can eat cotton, but not cloth, nor cotton bread or cotton pizza[^2]. FoodMothCottonSalad is not moth food. There are no high-quality moth foods. :(~~ – Based on a misunderstanding of how the new system works, not relevant.
* High-quality food are pizzas (but not their slices, which are bland), pies, cakes, cak (but not bread dog), and a few random things. It seems totally arbitrary and almost assigned accidentally rather than purposefully made to encourage good cooking
* In general, the distinction between Normal and High seems pretty arbitrary. Why is the FoodBurgerCrazy a high-quality food when it's made with a crayon and an emergency flare? Why is FoodBurgerEmpoweredRotten not toxic? Why are _all pizzas_ high-quality, but only some burgers? Why are no tacos high-quality?
* A few player-made foods are bland for reasons I don't understand.
    - Valid salad, which requires a high-effort meatball, unlike all the other salads. In fact, it contains doctor's delight – it should probably be a high-quality food!
    - Pea and tomato soups – why?
    - Bagels, regular and poppy. These should be Normal, imo.
* A bit of inconsistency around raw meats. Generally the full slab is Nasty or Toxin, but the corresponding cutlet is Bland. This inconsistency should be fixed. See the spreadsheet for details.

Overall, a bloody good effort. Mostly need to address a few misclassifications and stuff, maybe try to add a few more high-quality foods.

[^1]: On upstream, moths can eat ClothMade things too, but not on Frontier. Our commented-out moth digestion tags are out of date.
[^2]: These are ClothMade rather than MothFood, so aren't even in the spreadsheet. Cotton pizza _slices_ are moth-edible.

whatston3 commented:
You're right.  The quality of food is somewhat arbitrary (and at some level, it should be), but the goal is (and I think justly) to reward players for putting work into food, and an automated way to generate flavorol according to a set of rules would be good.  Two levels of food quality (normal/high) isn't great, three might be a sweet spot.  Simple recipes to make should be normal, a nice meal should be high quality, and difficult to make recipes should be gourmet (looking at you, non-citrus salads).

At the moment, the moth food tag only matters if it has RequiresSpecialDigestion.  While we could add a separate bonus to moths for eating moth food (and arguably a malus the other way - carnivore should also receive the same treatment), the benefit of "only you can eat this food" is fair enough.  Added this to the cotton bread and slice, since they were missing.

The toxin tag is in a weird spot.  I've been making a set of changes to the rework data - thank you for the spreadsheet.
The biggest change is that you have three-ish sets of qualities: "normal food", "goblin food", and "inedible crap that nobody should eat", with three steps in each.  With this set of changes, the rotten meat you get off of corpses should also be pretty much inedible.
https://docs.google.com/spreadsheets/d/1KIYgaokv8Mk5B7HY2ME691ryRdM-uqOQ955YtiKIVn0/edit?usp=sharing

I've removed rotten food off of expeditions in favor of food-related trash, and I've made a half set of changes for other things.  I'll mark this as awaiting changes for now, but I think it's still fair - **might need another pass, I'll aim to finish it up within the week**.  Bunch of oversights still remaining.

I've also added benefits for ruminants so cows get extra benefits for eating raw produce, I think this is probably fair, cows are really hungry as-is.

A few other tricks we could pull:
* Some species could eat brains and convert GreyMatter into something edible. (Protein, etc.)
* Vegetarian diets (for harpies? As a selectable trait?)

blackknight954 commented:
So as part of a note I had left myself earlier, is this PR going to adjust the timer until you get the Peckish and Starving debuffs if you eat high quality foods? If not that might be neat to look into,

whatston3 commented:
> So as part of a note I had left myself earlier, is this PR going to adjust the timer until you get the Peckish and Starving debuffs if you eat high quality foods? If not that might be neat to look into,

Not directly.  Higher quality foods (or, in general, foods well-suited to your diet) should give you more flavorol, and flavorol is being removed from base foods.  Depending on your species-specific digestion needs, foods (and only foods) turn into different chemicals before entering your stomach.  Flavorol as a mechanic is intended to keep you fed for longer, and gives you a movement speed bonus on top of that (15% off the top of my head), but doesn't affect the thresholds.

whatston3 commented:
I like the look of the imp food rebalance
getting that, a rewarming system, a food quality and/or digestions to generate flavorol, and combining that into one thing, probably the way

soon
